signer/src/stacks/api.rs

```diff
@@ -633,76 +625,120 @@ impl StacksClient {
+            break;
         }
-        _ => (),
+        // There are more blocks to fetch, so let's get them.
+        let mut tenure_blocks = stacks.get_tenure(block.header.parent_block_id).await?;
+        blocks.append(&mut tenure_blocks);
     }
-    result
+
+    Ok(blocks)
 }
 
-impl StacksInteract for StacksClient {
+/// A deserializer for Clarity's [`Value`] type that deserializes a hex-encoded
+/// string which was serialized using Clarity's consensus serialization format.
+fn clarity_value_deserializer<'de, D>(deserializer: D) -> Result<Value, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Value::try_deserialize_hex_untyped(&String::deserialize(deserializer)?)
+        .map_err(serde::de::Error::custom)
+}
+
+impl StacksInteract for ApiFallbackClient<StacksClient> {
     async fn get_current_signer_set(
-        &mut self,
+        &self,
         contract_principal: &StacksAddress,
     ) -> Result<Vec<PublicKey>, Error> {
-        // Make a request to the sbtc-registry contract to get the current
-        // signer set.
-        let result = StacksClient::get_data_var(
-            self,
-            contract_principal,
-            &ContractName::from("sbtc-registry"),
-            &ClarityName::from("current-signer-set"),
-        )
-        .await;
-
-        let checked_result = check_result(self, result)?;
-
-        match checked_result {
-            Value::Sequence(SequenceData::List(ListData { data, .. })) => {
-                let mut keys = Vec::new();
-                for item in data {
-                    if let Value::Sequence(SequenceData::Buffer(BuffData { data })) = item {
-                        keys.push(PublicKey::from_slice(&data)?);
-                    } else {
-                        return Err(Error::InvalidStacksResponse(
-                            "expected a buffer but got something else",
-                        ));
+        self.exec(|client, retry| async move {
+            // Make a request to the sbtc-registry contract to get the current
+            // signer set.
+            let result = StacksClient::get_data_var(
+                client,
+                contract_principal,
+                &ContractName::from("sbtc-registry"),
+                &ClarityName::from("current-signer-set"),
+            )
+            .await?;
+
+            match result {
+                Value::Sequence(SequenceData::List(ListData { data, .. })) => {
+                    let mut keys = Vec::new();
+                    for item in data {
+                        if let Value::Sequence(SequenceData::Buffer(BuffData { data })) = item {
+                            keys.push(PublicKey::from_slice(&data)?);
+                        } else {
+                            retry.abort();
+                            return Err(Error::InvalidStacksResponse(
+                                "expected a buffer but got something else",
+                            ));
+                        }
                     }
```

Comment:
I think this closure is a little too complex. Better to make it an actual function that is called in the closure and abort on the errors that are abort-able.

Reply:
This became obsolete after our recent discussion of moving actual logic back to the base clients, so now it's back to more how it was (with more comments).